src/DataGrid.tsx

```diff
@@ -368,43 +550,130 @@ function DataGrid<R, K extends keyof R, SR>({
+        mode: 'EDIT',
+        idx: selectedPosition.idx,
+        onKeyDown: handleKeyDown,
+        editorContainerProps: {
+          editorPortalTarget,
+          rowHeight,
+          scrollLeft,
+          scrollTop,
+          firstEditorKeyPress: selectedPosition.key,
+          onCommit: handleCommit,
+          onCommitCancel: () => setSelectedPosition(({ idx, rowIdx }) => ({ idx, rowIdx, mode: 'SELECT' }))
         }
+      };
+    }
+
+    return {
+      mode: 'SELECT',
+      idx: selectedPosition.idx,
+      onKeyDown: handleKeyDown,
+      dragHandleProps: enableCellDragAndDrop && isCellEditable(selectedPosition)
+        ? { onMouseDown: handleMouseDown, onDoubleClick: handleDoubleClick }
+        : undefined
+    };
+  }
+
+  function getRow(rowIdx: number) {
+    const row = rows[rowIdx];
+    let key: string | number = rowIdx;
+    let isRowSelected = false;
+    if (rowKey !== undefined) {
+      const rowId = row[rowKey];
+      isRowSelected = selectedRows?.has(rowId) ?? false;
+      if (typeof rowId === 'string' || typeof rowId === 'number') {
+        key = rowId;
       }
+    }
 
-      rowElements.push(
-        <RowRenderer
-          key={key}
-          rowIdx={rowIdx}
-          row={row}
-          viewportColumns={viewportColumns}
-          lastFrozenColumnIndex={lastFrozenColumnIndex}
-          eventBus={eventBus}
-          isRowSelected={isRowSelected}
-          onRowClick={onRowClick}
-          rowClass={rowClass}
-          top={rowIdx * rowHeight + totalHeaderHeight}
-        />
-      );
+    return (
+      <RowRenderer
+        key={key}
+        rowIdx={rowIdx}
+        row={row}
+        viewportColumns={viewportColumns}
+        lastFrozenColumnIndex={lastFrozenColumnIndex}
+        eventBus={eventBus}
+        isRowSelected={isRowSelected}
+        onRowClick={onRowClick}
+        rowClass={rowClass}
+        top={rowIdx * rowHeight + totalHeaderHeight}
+        copiedCellIdx={copiedPosition?.rowIdx === rowIdx ? copiedPosition.idx : undefined}
+        draggedOverCellIdx={getDraggedOverCellIdx(rowIdx)}
+        setDraggedOverRowIdx={isDragging ? setDraggedOverRowIdx : undefined}
+        selectedCellProps={getSelectedCellProps(rowIdx)}
+      />
+    );
+  }
+
+  function getViewportRows() {
+    const rowElements = [];
+
+    // Add the row containing the selected cell if not included in the vertical range
+    if (isCellWithinBounds(selectedPosition) && (selectedPosition.rowIdx < rowOverscanStartIdx || selectedPosition.rowIdx > rowOverscanEndIdx)) {
+      rowElements.push(getRow(selectedPosition.rowIdx));
+    }
+
+    for (let rowIdx = rowOverscanStartIdx; rowIdx <= rowOverscanEndIdx; rowIdx++) {
+      rowElements.push(getRow(rowIdx));
     }
 
     return rowElements;
   }
 
+  // Reset the positions if the current values are no longer valid. This can happen if a column or row is removed
+  if (selectedPosition.idx >= columns.length || selectedPosition.rowIdx >= rows.length) {
```

Comment:
Is it valid for resetting state in during render? Shouldn't we move these lines in an effect?

Reply:
AFAIK it's valid if it's the same component's state. If you're updating a parent component's state on the other hand React will give you a warning.

Reply:
That is correct. Check https://reactjs.org/docs/hooks-faq.html#how-do-i-implement-getderivedstatefromprops

Reply:
Also check https://github.com/facebook/react/pull/17099